src/JSInterop/Microsoft.JSInterop/src/IJSRuntime.cs

```diff
@@ -36,4 +52,26 @@ public interface IJSRuntime
     /// <param name="args">JSON-serializable arguments.</param>
     /// <returns>An instance of <typeparamref name="TValue"/> obtained by JSON-deserializing the return value.</returns>
     ValueTask<TValue> InvokeAsync<[DynamicallyAccessedMembers(JsonSerialized)] TValue>(string identifier, CancellationToken cancellationToken, object?[]? args);
+
+    /// <summary>
+    /// Invokes the specified JavaScript function asynchronously.
+    /// </summary>
+    /// <typeparam name="TValue">The JSON-serializable return type.</typeparam>
+    /// <param name="identifier">An identifier for the function to invoke. For example, the value <c>"someScope.someFunction"</c> will invoke the function <c>window.someScope.someFunction</c>.</param>
+    /// <param name="options">The <see cref="JsonSerializerOptions"/> to use for JSON serialization and deserialization.</param>
+    /// <param name="cancellationToken">
+    /// A cancellation token to signal the cancellation of the operation. Specifying this parameter will override any default cancellations such as due to timeouts
+    /// (<see cref="JSRuntime.DefaultAsyncTimeout"/>) from being applied.
+    /// </param>
+    /// <param name="args">JSON-serializable arguments.</param>
+    /// <returns>An instance of <typeparamref name="TValue"/> obtained by JSON-deserializing the return value.</returns>
+    ValueTask<TValue> InvokeAsync<[DynamicallyAccessedMembers(JsonSerialized)] TValue>(string identifier, JsonSerializerOptions options, CancellationToken cancellationToken, object?[]? args)
+        => throw new InvalidOperationException($"Supplying a custom {nameof(JsonSerializerOptions)} is not supported by the current JS runtime");
+
+    /// <summary>
+    /// Returns a copy of the current <see cref="JsonSerializerOptions"/> used for JSON serialization and deserialization.
+    /// </summary>
+    /// <returns>A copy of the <see cref="JsonSerializerOptions"/>.</returns>
+    JsonSerializerOptions CloneJsonSerializerOptions()
```

Comment:
Adding cloning APIs is not strictly necessary here. The JSON type itself exposes a copy constructor so you only need to expose the JSO being used as a property getter. Then the user would only need to write:
```C#
JsonSerializerOptions clone = new(runtime.JsonSerializerOptions);
```

Reply:
Isn't `JsonSerializerOptions` mutable? If so, exposing it would be very dangerous since this comes from a scoped DI service. Random 3rd-party NuGet packages could access the instance and mutate it, interfering with one another.

Reply:
It's mutable until first used whereby it [becomes frozen](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializeroptions.isreadonly?view=net-8.0). This can be triggered explicitly via the `options.MakeReadOnly()` method, which should ensure options are never returned by users in their mutable state.

Reply:
Cool, thanks for clarifying. That seems like a reasonable alternative then.

Reply:
We could definitely do this, it just introduces the question of when to call `MakeReadOnly()`. The `JSRuntime` abstract class can't do this in its constructor, because it already lets derived classes freely mutate the `JsonSerializerOptions`. So it would be up to each implementation to call `MakeReadOnly()` when it's done mutating the options. We can do this internally for each of our own platform-specific `JSRuntime` implementations, but anyone else extending `JSRuntime` will now suddenly be exposing their `JsonSerializerOptions` publicly.

If we're worried about this "breaking" customers by invalidating an assumption they might have made, we could make the exposure of `JsonSerializerOptions` opt-in via a new virtual property on `JSRuntime`, something like `protected virtual bool ExposeJsonSerializerOptions => false`. Or we could add a `protected virtual JsonSerializerOptions GetJsonSerializerOptions()`, so that concrete types can lazily mark the `JsonSerializerOptions` as read-only on first access, if they want.

But I'm going to move forward assuming that we're okay with slightly impacting the somewhat rare scenario of extending `JSRuntime`. We have the option of making an announcement about it.

Reply:
> But I'm going to move forward assuming that we're okay with slightly impacting the somewhat rare scenario of extending JSRuntime. Types extending JSRuntime should just call MakeReadOnly() after the JsonSerializerOptions are done being configured. We have the option of making an announcement about it.

Sounds fine to me. 99% of the time, people are using the built-in JS runtime type, and the only other case we know of (I think) is unit/integration tests. If people want to do random stuff in their tests they can do so. As long as 3rd-party packages don't mutate the options used in real app code I think it's fine.